### ISSUE TITLE
Support combined cluster count and cluster weak lensing

### ISSUE THREAD

**Issue by NiallMac** [joezuntz/2point#2]
We want to extend the library/file format to be applicable to combined cluster count and cluster weak lensing. The cluster weak lensing signal is a 2-point function, the SpectrumMeasurement class should work, possibly with some small extensions if we are using DeltaSigma(R) rather than gamma_t(theta)). There are a few extra considerations:

i) Boost factor - a value per tangential shear datapoint. The boost factors have a covariance matrix. Is there covariance between the boost factors and the raw measurements?

ii) Clusters are split into richness bins as well as redshift bins. This does not necessarily require generalizing the SpectrumMeasurement class. We could either 

a. Have a separate SpectrumMeasurement instance (i.e. separate extensions at the file level) per richness bin (each containing all the z bin combinations for that richness bin)...

b. Hold all (richness, redshift) bins in one SpectrumMeasurement instance (i.e. the same extension at the file level) e.g. ordered (R_i=richness bin i, zc_i=cluster redshift bin j, zs_k=source redshift bin k):
R_1, zc_1, zs_1
R_1, zc_1, zs_2
...
R_1, zc_2, zs_1
R_1, zc_2, zs_2
...
R_2, zc_1, zs_1
R_2, zc_1, zs_2
etc.
The 'bin1' index would then be R_i * (# of cluster z bins) + zc_j (and the bin2 index would just be the source redshift bin index k as usual). 
For this solution, we would probably still want to generalize (make a child class) of SpectrumMeasurement which can translate between bin1 and R_i, zc_j.

iii) We want to include count information. This is just a number (per area?) per richness, cluster z bin. It also needs an accompanying row,column in the covariance matrix.

**Comment by danielgruen** [comment]
That's great! Some comments ...

2018-02-09 10:22 GMT-08:00 NiallMac <notifications@github.com>:

> We want to extend the library/file format to be applicable to combined
> cluster count and cluster weak lensing. The cluster weak lensing signal is
> a 2-point function, the SpectrumMeasurement class should work, possibly
> with some small extensions if we are using DeltaSigma(R) rather than
> gamma_t(theta)). There are a few extra considerations:
>
I think you can rely on us using gamma_t(theta) for a fixed source redshift
distribution per source redshift bin (not different ones for each
radial/richness/lens redshift bins etc.)

> i) Boost factor - a value per tangential shear datapoint. The boost
> factors have a covariance matrix. Is there covariance between the boost
> factors and the raw measurements?
>

I don't think there is a cross-covariance.

Note that Eduardo, Tamas and I are still discussing whether we need the
boost factor data points or can do the correction at the data vector level.
I am thinking the former, but maybe I am thinking too complicated.

> ii) Clusters are split into richness bins as well as redshift bins. This
> does not necessarily require generalizing the SpectrumMeasurement class. We
> could either
>
> a. Have a separate SpectrumMeasurement instance (i.e. separate extensions
> at the file level) per richness bin (each containing all the z bin
> combinations for that richness bin)...
>
> b. Hold all (richness, redshift) bins in one SpectrumMeasurement instance
> (i.e. the same extension at the file level) e.g. ordered (R_i=richness bin
> i, zc_i=cluster redshift bin j, zs_k=source redshift bin k):
> R_1, zc_1, zs_1
> R_1, zc_1, zs_2
> ...
> R_1, zc_2, zs_1
> R_1, zc_2, zs_2
> ...
> R_2, zc_1, zs_1
> R_2, zc_1, zs_2
> etc.
> The 'bin1' index would then be R_i * (# of cluster z bins) + zc_j (and the
> bin2 index would just be the source redshift bin index k as usual).
> For this solution, we would probably still want to generalize (make a
> child class) of SpectrumMeasurement which can translate between bin1 and
> R_i, zc_j.
>
Option b sounds better to me, and also maps well to the cosmolike way of
handling data.


> iii) We want to include count information. This is just a number (per
> area?) per richness, cluster z bin. It also needs an accompanying
> row,column in the covariance matrix.
>
I guess it's really the raw count at the data vector level. Where is the
meta data about the other data vectors that's required for things like
covariances (source density, survey area, etc.)? The information on area
should be handled the same way.

**Comment by NiallMac** [comment]
@danielgruen Do we need to save n(Lambda) (as in the distribution, not just the total) for each redshift/richness bin? I would have thought this would be required to predict P(M) for that bin e.g. P(M) = \int dlambda P(M | lambda) n(lambda) or something...

**Comment by NiallMac** [comment]
Apparently this isn't necessary - next question is how to include the redshift selection function information. The redshift selection function for a given cluster bin is just a top-hat in cluster photo-z. However, information from the catalogs (and I think this file should contain all information required from the catalogs) is required to generate P(z_true | photo-z) for the likelihood calculation. How should we store this information? Couple of options:

i) For each cluster bin, store arrays of finely spaced z, mean(sigma_z) and std(sigma_z) (where sigma_z is the RedMapper reported photo-z error, and the mean/std is over all clusters in some finely spaced z bin.

ii) Matteo uses a polynomial fit to sigma_z(z) for each lambda bin - we could store an object that can store the polynomial coefficients.

Both of these options could be accompanied by a function that returns sigma_z(z, cluster bin). One advantage of (ii) is that it will be faster for repeated evaluations e.g. in integrals. But of course one could also fit a polynomial to the arrays stored in option (i).

Thoughts?

**Comment by danielgruen** [comment]
Yeah, that sounds reasonable. I guess the numbers make the ordering of the
polynomial coefficients pretty unambiguous ...
c[0]*z**N+c[1]*z**(N-1)+c[N-1]



2018-02-10 12:10 GMT-08:00 NiallMac <notifications@github.com>:

**Comment by NiallMac** [comment]
Yeh, that is what I went with in the end. There's a get_sigma_z function that at least removes ambiguity for users of an already made file. Hmm...but yeh the ordering on input is fairly ambiguous, I guess this just needs to be  documented this very clearly...

**Comment by danielgruen** [comment]
FWIW, I've updated the mock-ish Y1 files to include a mock-ish polynomial
sigma_z at https://www.slac.stanford.edu/~dgruen/lighthouse/


2018-02-14 19:08 GMT-08:00 NiallMac <notifications@github.com>:

**Comment by chto** [comment]
@danielgruen @NiallMac @joezuntz  
Just check whether the stuff discussed above has been implemented.  We are deciding what format we want to use to release the large-scale cluster key project data vector.

**Comment by joezuntz** [comment]
I don't think we ever did this, no.  In general I'm gradually transitioning to [SACC](https://github.com/LSSTDESC/sacc), so if it would require new code either way then I'd suggest putting it in that form. There is a sacc likelihood in cosmosis that we can extend if needed.

**Comment by chto** [comment]
Thank you, @joezuntz, for the quick reply!
